### PR TITLE
feat: add support for nested substitutions

### PR DESCRIPTION
**Pull request by Bizordec**
Adds support for nested substitutions in MyST parser configuration. Nested dictionary structures are automatically flattened using dot notation.

Example:
```python
myst_substitutions = {
    "a": "value_a",
    "b": {"c": "value_b_c", "d": {"e": "value_b_d_e"}}
}
# Flattens to: {"a": "value_a", "b.c": "value_b_c", "b.d.e": "value_b_d_e"}
```

### CODE REVIEW

**Comment by adamtheturtle** [comment]
Thank you @Bizordec for contributing. I'd like to understand more.

I'm not super familiar with MyST, and I can't find any reference to this dot notation being used in MyST tooling. Am I missing something?

If this is new - why MyST only?

If this is very custom, rather than shipping this feature, could a user not have a flattener in their `conf.py` instead?

**Comment by Bizordec** [comment]
> I can't find any reference to this dot notation being used in MyST tooling.

That dot notation isn't documented as explicit syntax. However, since MyST uses Jinja2 for substitutions, nested structures in conf.py or frontmatter can be accessed via dot notation (e.g., `{{ nested_dict.key1 }}`). 
See: https://github.com/executablebooks/MyST-Parser/commit/3a6bbb1637ad0eadea0908c7bab28c2e7aa714f3#diff-189f46bc22888254cc0b6760e398aa446fa201f5d63b45d72c1e991a13e28cae

> If this is new - why MyST only?

Because reST substitutions are defined inline using `.. |name| replace:: replacement *text*`, not in conf.py. There's no equivalent nested structure to flatten for reST.

> If this is very custom, rather than shipping this feature, could a user not have a flattener in their conf.py instead?

They could, but it would require maintaining flat keys manually:
```python
myst_substitutions = {
    "a": "value_a",
    "b_c": "value_b_c",
    "b_c_d_e": "value_b_d_e",
}
```

Instead of the more maintainable nested structure:

```python
myst_substitutions = {
    "a": "value_a",
    "b": {
        "c": "value_b_c",
        "d": {
            "e": "value_b_d_e",
        },
    },
}
```

---

I've also added support for lists:
`{'items': [{'name': 'a'}]} -> {'items.0.name': 'a'}`

and removed beartype for `_flatten_substitutions` function, because it can't resolve recursive type:
```
beartype.roar.BeartypeCallHintForwardRefException: Forward reference  PEP 604 new union str | int | float | list['SubstitutionValue'] | dict[str, 'SubstitutionValue'] uncheckable at runtime, as new union child type 3 list['SubstitutionValue'] not class.
```

**Comment by adamtheturtle** [comment]
Thanks @Bizordec, that's a great explanation — the Jinja2 dot-notation link makes the rationale clear, and explains why this is MyST-only. I'm happy to go ahead with the feature.

I've pushed a few commits to your branch tidying things up to my preferences (one commit per change):

- Made `prefix` a required keyword-only argument on `_flatten_substitutions` rather than defaulting it to `""`.
- Removed the `# type: ignore` comments by fixing the underlying issues: `enumerate` and `str` are now called with keyword arguments (`enumerate(iterable=...)`, `str(object=...)`), which is what the `mypy_strict_kwargs` plugin expects, so nothing needs suppressing. (Note `pyright` and `ty` are configured to ignore `# type: ignore` comments here, so suppressing wouldn't have worked for them anyway.)
- Replaced the `isinstance` if/elif/else chain with a `match` statement (our minimum supported Python is 3.11, so this is fine).

All of mypy, pyright, ty and pyrefly pass, and your two new tests still pass.

What's still left before this can merge:

- **Docs**: the nested/list substitution behaviour needs documenting (including the dot notation and list-index notation).
- **Tests**: a little more coverage would be good, e.g. mixed nesting and any edge cases.

**Comment by Bizordec** [comment]
Thank you for the feedback, done.

**Comment by adamtheturtle** [comment]
Thank you! Can there be overlaps, e.g. if a key includes `.`? Do we match MyST for this case? Is this tested?

File: src/sphinx_substitution_extensions/__init__.py [hunk 1]
```diff
@@ -34,6 +34,51 @@
...
+    str
+    | int
+    | float
+    | list["SubstitutionValue"]
+    | dict[str, "SubstitutionValue"]
+)
+Substitutions: TypeAlias = dict[str, SubstitutionValue]
+
+
+def _flatten_substitutions(
```
Comment (adamtheturtle) [inline src/sphinx_substitution_extensions/__init__.py:71]: Please add a comment about beartype

File: src/sphinx_substitution_extensions/__init__.py [hunk 2]
```diff
@@ -34,6 +34,51 @@
...
+    *,
+    substitutions: Substitutions,
+    prefix: str,
```
Comment (adamtheturtle) [inline src/sphinx_substitution_extensions/__init__.py]: Is the prefix parameter ever used? If not, remove it

Comment (Bizordec) [reply src/sphinx_substitution_extensions/__init__.py]: It is used as the initial value for the stack:
`stack: list[tuple[str, SubstitutionValue]] = [(prefix, substitutions)]`

I can remove `prefix` and just init stack like this:
`stack: list[tuple[str, SubstitutionValue]] = [("", substitutions)]`

Comment (adamtheturtle) [reply src/sphinx_substitution_extensions/__init__.py]: Whichever you think is best

File: tests/test_substitution_extensions.py [hunk 3]
```diff
@@ -1386,6 +1386,714 @@ def test_myst_substitutions(
...
+        ).read_text()
+        assert content_html == expected_content_html
+
+    @staticmethod
+    def test_myst_empty_containers(
+        *,
+        tmp_path: Path,
+        make_app: Callable[..., SphinxTestApp],
+    ) -> None:
+        """MyST empty containers (lists and dicts) do not create keys
```
Comment (adamtheturtle) [inline tests/test_substitution_extensions.py]:
```suggestion
        """MyST empty containers (lists and dictionaries) do not create keys
```

This will avoid the spellchecker issues

File: src/sphinx_substitution_extensions/__init__.py (hunk 1 above)
Comment (Bizordec) [reply src/sphinx_substitution_extensions/__init__.py:71]: Done

File: src/sphinx_substitution_extensions/__init__.py (hunk 2 above)
Comment (Bizordec) [reply src/sphinx_substitution_extensions/__init__.py]: Done

**Comment by Bizordec** [comment]
> Thank you! Can there be overlaps, e.g. if a key includes `.`? Do we match MyST for this case? Is this tested?

Good point, it will overlap. I think we should prohibit dots in key names and add a note to the docs.

**Comment by adamtheturtle** [comment]
> Good point, it will overlap. I think we should prohibit dots in key names and add a note to the docs.

Ok, please do that

File: src/sphinx_substitution_extensions/__init__.py [hunk 4]
```diff
@@ -44,18 +45,68 @@
 Substitutions: TypeAlias = dict[str, SubstitutionValue]
 
 
+@beartype
+def _validate_substitution_key(*, key: str) -> None:
```
Comment (adamtheturtle) [inline src/sphinx_substitution_extensions/__init__.py:49]: Does this match what happens with MyST generally? If not, how do they handle the overlap case?

File: tests/test_substitution_extensions.py [hunk 5]
```diff
@@ -2095,6 +2097,122 @@ def test_myst_invalid_substitution_access(
         for text in expected_text_in_html:
             assert text in content_html
 
+    @staticmethod
+    def test_myst_substitution_key_with_dot_raises_error(
```
Comment (adamtheturtle) [inline tests/test_substitution_extensions.py:2101]: Please add a test that there is no such error for rST, as I believe some code refactoring could easily introduce that.

File: README.rst [hunk 6]
```diff
@@ -237,6 +237,8 @@ Nested substitutions
 
 ``myst_substitutions`` supports nested dictionaries and lists, which are flattened using dot notation.
 
+**Important:** Substitution keys cannot contain dots (``.``), as dots are reserved for nested access notation. For example, ``{"key.with.dots": "value"}`` is not allowed, but ``{"key": {"with": {"dots": "value"}}}`` is valid and accessed as ``|key.with.dots|``.
```
Comment (adamtheturtle) [inline README.rst]: We use one-sentence-per-line in documentation source for easier editing.

Comment (adamtheturtle) [inline README.rst]: Change "is not allowed" to "raises an exception" (or similar).

Comment (adamtheturtle) [inline README.rst]: Do you think that everything from "but" is valuable? It seems to repeat what is below and I suggest removing it.

File: src/sphinx_substitution_extensions/__init__.py (hunk 4 above)
Comment (adamtheturtle) [reply src/sphinx_substitution_extensions/__init__.py:49]: MyST itself never flattens, so it has no overlap to resolve. `render_substitution` just builds `variable_context = {**self.md_config.substitutions}` and renders `{{ <ref> }}` through Jinja2 (`StrictUndefined`). Nesting is therefore resolved natively by Jinja2 attribute/item lookup (`{{ a.b.c }}`), with no flat key ever created. A literal-dot key like `{"key.with.dots": "value"}` isn't a collision there either — `{{ key.with.dots }}` is just attribute access that fails into a substitution warning, and the value is only reachable via subscript, which the `|name|` syntax can't express anyway.

Our case is different because `|a.b.c|` is plain string replacement, not Jinja2, so flattening to a single string-keyed dict is the only way to support nesting — and that flattening is precisely what creates the overlap. So prohibiting dots is our own rule rather than something MyST does; it's the minimal way to keep flattening unambiguous. (Verified against the installed `myst_parser.mdit_to_docutils.base.render_substitution`.)

File: tests/test_substitution_extensions.py (hunk 5 above)
Comment (adamtheturtle) [reply tests/test_substitution_extensions.py:2101]: Added `test_rst_substitution_key_with_dot_does_not_raise_error` in 857a202 — it defines `.. |key.with.dots| replace::` and asserts the build succeeds, so a refactor that leaked the dot validation into the reST path would fail here.

File: README.rst (hunk 6 above)
Comment (adamtheturtle) [reply README.rst]: Done in 857a202 — reflowed to one sentence per line.

Comment (adamtheturtle) [reply README.rst]: Done in 857a202 — now reads "raises an exception".

Comment (adamtheturtle) [reply README.rst]: Done in 857a202 — removed the trailing "but ..." clause.

File: src/sphinx_substitution_extensions/__init__.py (hunk 4 above)
Comment (Bizordec) [reply src/sphinx_substitution_extensions/__init__.py:49]: Yeah, I think it would still be better to just prohibit dots in keys as our own rule.

Here is my testing of substitutions usage in MyST if you interested:

Снимок экрана 2026-06-17 в 10 58 31

**Comment by adamtheturtle** [comment]
This has now been released. Thank you for iterating with me @Bizordec !
